components/builder-depot/src/upstream.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bldr_core::logger::Logger;
+use hab_core::package::{Identifiable, PackageIdent};
+use hab_net::socket::DEFAULT_CONTEXT;
+use iron::typemap::Key;
+use protobuf::{parse_from_bytes, Message};
+use std::collections::{HashSet, VecDeque};
+use std::path::PathBuf;
+use std::sync::mpsc;
+use std::thread::{self, JoinHandle};
+use std::time::{Duration, Instant};
+
+use zmq;
+
+use config::Config;
+use error::{Error, Result};
+use protocol::originsrv::OriginPackageIdent;
+
+use depot_client::Client as DepotClient;
+use server::download_package_from_upstream_depot;
+
+const UPSTREAM_MGR_ADDR: &'static str = "inproc://upstream";
+const DEFAULT_POLL_TIMEOUT_MS: u64 = 60_000; // 60 secs
+
+pub struct UpstreamClient;
+
+pub struct UpstreamCli;
+
+impl Key for UpstreamCli {
+    type Value = UpstreamClient;
+}
+
+impl UpstreamClient {
+    pub fn refresh(&self, ident: &OriginPackageIdent) -> Result<()> {
+        // TODO: Use a per-thread socket when we move to a post-Iron framework
+        let socket = (**DEFAULT_CONTEXT).as_mut().socket(zmq::DEALER).unwrap();
+        socket.connect(UPSTREAM_MGR_ADDR).map_err(Error::Zmq)?;
+        socket
+            .send(&ident.write_to_bytes().unwrap(), 0)
+            .map_err(Error::Zmq)?;
+        Ok(())
+    }
+}
+
+impl Default for UpstreamClient {
+    fn default() -> UpstreamClient {
+        UpstreamClient {}
+    }
+}
+
+pub struct UpstreamMgr {
+    config: Config,
+    depot_client: Option<DepotClient>,
+    upstream_mgr_sock: zmq::Socket,
+    want_origins: HashSet<String>,
+    logger: Logger,
+    msg: zmq::Message,
+}
+
+impl UpstreamMgr {
+    pub fn new(cfg: &Config) -> Result<Self> {
+        let upstream_mgr_sock = (**DEFAULT_CONTEXT)
+            .as_mut()
+            .socket(zmq::DEALER)
+            .map_err(Error::Zmq)?;
+
+        let depot_client = if let Some(ref upstream_depot) = cfg.upstream_depot {
+            Some(DepotClient::new(upstream_depot, "builder-upstream", "0.0.0", None).unwrap())
+        } else {
+            None
+        };
+
+        let log_path = cfg.log_dir.clone();
+        let mut logger = Logger::init(PathBuf::from(log_path), "builder-upstream.log");
+
+        let want_origins: HashSet<String> =
+            cfg.upstream_origins.iter().map(|s| s.to_owned()).collect();
```

Comment:
Is the `to_owned()` necessary here?  Isn't `cfg.upstream_origins` already a `Vec<String>`?

Reply:
In fact, since you're just converting a `Vec` to a `HashSet`, you could probably do that with something like:

`let want_origins: HashSet<String> = HashSet::from_iter(cfg.upstream_origins);`

Reply:
👍 will try that

Reply:
ah no, its not that simple - the ownership does need to change as we cant hand the ownership of the strings from the borrowed config to the hashset.  so the from_iter fails with a borrow checker error.

Reply:
Ah, I forgot about the borrowed config.  Oh well 😁

Reply:
one can run, but not hide from the borrow checker :)